### PR TITLE
DO NOT MERGE: Hooked up space calculator

### PR DESCRIPTION
**Pull request by escowles**
PR for code review only, do not merge.

### CODE REVIEW

**Comment by escowles** [comment]
I've rebased this (without squashing)

File: frontend/app/controllers/top_containers_controller.rb [hunk 1]
```diff
@@ -169,6 +169,24 @@ def update_barcodes
...
+    update_uris = params[:update_uris]
+    location_data = {}
+    update_uris.map{|uri| location_data[uri] = params[uri].blank? ? nil : params[uri]['ref']}
+
+    post_uri = "#{JSONModel::HTTP.backend_url}/repositories/#{session[:repo_id]}/top_containers/bulk/locations"
+
+    response = JSONModel::HTTP::post_json(URI(post_uri), location_data.to_json)
+    result = ASUtils.json_parse(response.body)
+
+    if response.code =~ /^4/
```
Comment (escowles) [inline frontend/app/controllers/top_containers_controller.rb]: Handling 5xx responses here would probably be a good idea too.

File: scripts/load_location_data.rb [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+PLACES = java.util.TimeZone.getAvailableIDs.map {|z| z.to_s.split('/')[1]}.compact.reject {|p| p =~ /[A-Z]{3}/}.map {|p| p.sub('_', ' ')}.uniq
+
+# Suffixes
+LOCATION_SUFFIXES = ['Square', 'Gardens', 'Library', 'Museum', 'Warehouse', 'Treehouse', 'Bunker']
+
+
+require 'jsonmodel'
+require 'securerandom'
+
+JSONModel::init(:client_mode => true, :url => 'http://localhost:4567')
```
Comment (escowles) [inline scripts/load_location_data.rb]: Would it be useful to make the URL here configurable?  Do people run on separate machines or use different port numbers?

File: scripts/load_location_data.rb [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+                                                               ]).save
+
+        puts "Created top container with ID: #{top_container_id}"
+      end
+    end
+  end
+end
+
+
+def main(username = 'admin', password = 'admin')
```
Comment (escowles) [inline scripts/load_location_data.rb]: Again, might be useful to make this configurable if people are not using the standard account in their dev/test setup for whatever reason.

File: backend/spec/model_space_calculator_spec.rb [hunk 4]
```diff
@@ -0,0 +1,176 @@
+require 'spec_helper'
```
Comment (escowles) [inline backend/spec/model_space_calculator_spec.rb]: More of a general comment: this is a very small amount of tests being added compared to the size of the overall diff (~200 lines of tests against ~3000 lines of code).  Adding a controller or feature test exercising some of this functionality is probably the most efficient way to test this, given the amount of frontend view and javascript updates.

**Comment by escowles** [comment]
Not specific to this PR: there are a ton of different Travis jobs, and all of the `travis:test` jobs are failing.  It looks like bad tests to me, e.g.: https://travis-ci.org/archivesspace/archivesspace/jobs/124158071#L2482 .  Some, but not all, of the `travis:selenium:staff` jobs are also failing, e.g.: https://travis-ci.org/archivesspace/archivesspace/jobs/124158084#L851

Fixing these tests and getting the test suite back to green would be a great thing to do, and probably not a huge task.

File: backend/app/model/space_calculator.rb [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+      while packer.add_container(@container_profile.name,
+                                 Dimensions.new(@container_profile.width,
+                                                @container_profile.height,
+                                                @container_profile.depth,
+                                                @container_profile.dimension_units),
+                                 @container_profile.stacking_limit)
+        count += 1
+
+        # peace of mind
+        break if count > MAX_CONTAINERS_TO_PACK
```
Comment (escowles) [inline backend/app/model/space_calculator.rb]: I am a little suspicious of this strategy, but this should at least be configurable so people can set their own limits for this.

**Comment by escowles** [comment]
Overall, this looks like very high quality code and I have only a handful of comments.  Lack of test coverage for the new code, and failing Travis builds are both red flags for me though.

**Comment by mark-cooper** [comment]
Just an observation, haven't dug into anything. Ran tests locally (`./build/run backend:test` specifically), and got failure:

```
...
     [java] 1 deprecation warning total
     [java] 
     [java] Finished in 10 minutes 57 seconds (files took 42.66 seconds to load)
     [java] 1392 examples, 4 failures, 1 pending
     [java] 
     [java] Failed examples:
     [java] 
     [java] rspec ./spec/model_extent_calculator_spec.rb:82 # Extent Calculator model deals with large resources
     [java] rspec ./spec/model_extent_calculator_spec.rb:91 # Extent Calculator model doesn't mind different kinds of containers
     [java] rspec ./spec/model_extent_calculator_spec.rb:117 # Extent Calculator model can calculate extent for subtrees
     [java] rspec ./spec/model_extent_calculator_spec.rb:130 # Extent Calculator model can provide a hash rendering of itself
     [java] 
     [java] Randomized with seed 1
     [java] 
     [java] 
     [java] Deprecation Warnings:
     [java] 
     [java] Using `stub` from rspec-mocks' old `:should` syntax without explicitly enabling the syntax is deprecated. Use the new `:expect` syntax or explicitly enable `:should` instead. Called from /home/mcooper/Projects/archivesspace/archivesspace/backend/spec/model_session_spec.rb:35:in `(root)'.
     [java] 
     [java] 
     [java] If you need more of the backtrace for any of these deprecations to
     [java] identify where to make the necessary changes, you can configure
     [java] `config.raise_errors_for_deprecations!`, and it will turn the
     [java] deprecation warnings into errors, giving you the full backtrace.

BUILD FAILED
/home/mcooper/Projects/archivesspace/archivesspace/build/build.xml:172: Java returned: 1

Total time: 11 minutes 43 seconds
```

**Comment by marktriggs** [comment]
Oops, failing tests are probably my fault, as I rebased from archivesspace/master for the purposes of the code review but didn't rerun them.  I'll figure out what's going on and fix them up as needed.

**Comment by marktriggs** [comment]
Ah actually, the failing tests are the *extent* calculator, not the new space calculator.  Too many calculators.

Here's the stack trace, for reference:

     [java]   4) Extent Calculator model can provide a hash rendering of itself
     [java]      Failure/Error: boxes = create_containers(bigbox_profile, 10)
     [java]      NoMethodError:
     [java]        undefined method `uri' for 10:Fixnum
     [java]      # ./spec/model_space_calculator_spec.rb:37:in `create_containers'
     [java]      # org/jruby/RubyFixnum.java:280:in `times'
     [java]      # ./spec/model_space_calculator_spec.rb:36:in `create_containers'
     [java]      # ./spec/model_extent_calculator_spec.rb:133:in `(root)'
     [java]      # org/jruby/RubyBasicObject.java:1623:in `instance_exec'
     [java]      # /mnt/ssd/archivesspace/build/gems/gems/rspec-core-3.3.2/lib/rspec/core/example.rb:206:in `run'
     [java]      # /mnt/ssd/archivesspace/build/gems/gems/rspec-core-3.3.2/lib/rspec/core/example.rb:430:in `with_around_and_singleton_context_hooks'
     [java]      # /mnt/ssd/archivesspace/build/gems/gems/rspec-core-3.3.2/lib/rspec/core/example.rb:388:in `with_around_example_hooks'
     [java]      # /mnt/ssd/archivesspace/build/gems/gems/rspec-core-3.3.2/lib/rspec/core/hooks.rb:478:in `run'
     [java]      # /mnt/ssd/archivesspace/build/gems/gems/rspec-core-3.3.2/lib/rspec/core/hooks.rb:618:in `run_around_example_hooks_for'
     [java]      # org/jruby/RubyProc.java:290:in `call'
     [java]      # org/jruby/RubyProc.java:271:in `call'

100 points for anyone who just monkey patches a `uri` method onto Fixnum ;)

**Comment by mark-cooper** [comment]
My mistake. Running just the new tests ... they pass =)

I gave the features a test drive as outlined in the specification doc and almost everything I tried checked out. Nice. The only (minor) "problem" has been with the space calculator ... if you click "Check for Space" before selecting a Floor or Room it displays "Loading" but won't actually do anything until you select something for those fields. Perhaps the fields should be required and / or the "Check" button disabled until all the fields are populated? A small thing but probably helpful. Also, in some cases, even with all the fields populated the "Loading" message appears without returning results. I'm sure there's a good underlying reason, but the UI could be more user friendly.

To replicate I was at:
http://aspace.hudmol.com:7100/container_profiles/4 (repo1)
Space Calculator
20 Cooper Square
Basement
Stacks
Click Check for Space
Result: Loading ... (forever)

**Comment by escowles** [comment]
Yes, the test failures may be unrelated to the changes in this PR.  But they are passing in master:

https://travis-ci.org/archivesspace/archivesspace/builds/122847991

So it would be good to figure out what's changed and get them passing again before merging this.

**Comment by marktriggs** [comment]
Ah!  I just noticed both `model_space_calculator_spec.rb` and `model_extent_calculator_spec.rb` have defined a method called `create_containers`, so one's probably clobbering the other and causing the tests to fail.

I think the easiest fix is if I just move the space calculator's helper methods into the `describe` block, which should limit their scope.  If that fixes it, I might do the same for the extent calculator too--just to avoid future such issues.

**Comment by cfitz** [comment]
Hi, 

quick note about travis...it's pretty flaky, especially on the selenium test. We had to break out the tests into so many task because the suite was running into travis' time limit. we also have a bunch of problems with consistently getting things to pass...usually I have to restart a few of the failures and they'll pass.

Running some coverage reports now and will look at them shortly..

**Comment by cfitz** [comment]
Hey, sorry realized I forgot to look back on this...

Yes, everything looks good to me, but maybe some more additional rspec tests ( coverage was low in some spots) and Selenium tests.
I have been slowly chipping away at trying to get Travis to be more consistent, which is always a battle for some reason. Will take a few more wacks at it during my last week here....

**Comment by mark-cooper** [comment]
@marktriggs @escowles offered this feedback last week:

"I see about 2 dozen commits since I last looked at the code.  I see some test coverage has been added, which is good.  But I haven't really looked at the updates otherwise, other than to notice that the handful of hard-coded URLs and passwords in the PR are still there."

Do you anticipate having an update soon? My understanding is that this work is intended for the 1.5.0 release so it would be good to have it incorporated into an RC3 if that sounds reasonable to you?

**Comment by jambun** [comment]
Hi @mark-cooper - The NYU folks are doing final acceptance testing. We don't know how long that will take.

On the hardcoded urls and passwords - that is just a script for loading test data, so we don't think there is a need to make them configurable. The sort of person likely to be running that script can "configure" it by editing the file. If you all still don't like it, we could just remove the file. Let us know!

Cheers,
James

@marktriggs @escowles

File: backend/app/model/space_calculator.rb (hunk 5 above)
Comment (jambun) [inline backend/app/model/space_calculator.rb]: The line is really just a paranoid defense against an improbable infinite loop.

**Comment by mark-cooper** [comment]
Closing. Pull request is here:

https://github.com/archivesspace/archivesspace/pull/467

**Pull request closed without merging**